Objects/obmalloc.c

```diff
@@ -2663,9 +2683,71 @@ _PyDebugAllocatorStats(FILE *out,
     (void)printone(out, buf2, num_blocks * sizeof_block);
 }
 
+int _PyMem_init_obmalloc(PyInterpreterState *interp, _PyRuntimeState *runtime)
+{
+#ifdef WITH_PYMALLOC
+    /* Initialize obmalloc, but only for subinterpreters,
+       since the main interpreter is initialized statically. */
+    if (interp == &runtime->_main_interpreter
+            || (interp->feature_flags & Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
+        interp->obmalloc = &obmalloc_state_main;
+        interp->obmalloc->heap_allocated = false;
+    } else {
+        interp->obmalloc = PyMem_RawCalloc(1, sizeof(struct _obmalloc_state));
+        if (interp->obmalloc == NULL) {
+            return 0;
+        }
+        interp->obmalloc->heap_allocated = true;
+    }
```

Comment:
Alternately, we don't actually need to set `interp->obmalloc` for the main interpreter (and returning 0 for success):

```suggestion
    struct _obmalloc_state *state;
    if (_Py_IsMainInterpreter(interp)) {
        assert(interp->obmalloc == NULL);
        state = &obmalloc_state_main;
        state->heap_allocated = false;
    }
    else if (_PyInterpreterState_HasFeature(interp, Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
        assert(interp->obmalloc == NULL);
        // It was already initialized for the main interpreter.
        return 0;
    }
    else {
        state = PyMem_RawCalloc(1, sizeof(struct _obmalloc_state));
        if (state == NULL) {
            return -1;
        }
        state->heap_allocated = true;
        interp->obmalloc = state;
    }
```